1/backend/src/resolvers.js

```diff
@@ -0,0 +1,86 @@
+const jwt = require('jsonwebtoken');
+const fs = require('fs');
+
+data = require('./data');
+
+let SECRET_KEY = fs.readFileSync('./src/key/secret.key', 'utf8');
+
+const resolvers = {
+    Query: {
+        todos: () => JSON.parse(JSON.stringify(data.todos))
+    },
+    Mutation: {
+        addTodo: (parent, args, context) => {
+            if(args.message != "" && args.message != null && context.token) {
+                let newTodo = {
+                    id: findNextId(),
+                    message: args.message,
+                    completed: false,
+                };
+                data.todos.push(newTodo);
+                return newTodo;
+            }
+            return;
+        },
+        deleteTodo: (parent, args, context) => {
+            if (args.id != null && context.token){
+                let index = data.todos.findIndex(td => td.id === parseInt(args.id));
+                if (index > 0){
+                    data.todos.splice( index, 1)
+                    return true
+                }
+                return false
+            }
+        },
+        editTodo: (parent, args, context) => {
+            if (args.id != null && args.message != null && context.token){
+                tod = data.todos.find(td => td.id === parseInt(args.id))
+                if (tod){
+                    tod.message = args.message;
+                    return tod;
+                }
+                return
+            }
+
+        },
+        finishTodo: (parent, args, context) => {
+            if (args.id != null && context.token){
+                let tod = data.todos.find(td => td.id === parseInt(args.id));
+                if (tod){
+                    //Reverse status of item
+                    tod.completed = !tod.completed;
+                    return tod
+                }
+                return
+            }
+        },
+        login: (parent, args) => {
+            if (args.usr != null && args.pwd != null){
+                let currentUsr = data.user.find(usr => usr.login === args.usr)
+                if (currentUsr != null && currentUsr.password === args.pwd){
+                    const token = jwt.sign(
+                        { name: currentUsr.name, id: currentUsr.id },
+                        SECRET_KEY,
+                        { expiresIn: '1d' }
+                    )
+                    return token
+                }
+                return currentUsr.name
```

Comment:
What? So you randomly get the username returned if the password is incorrect?
1. The client might confuse the returned username with a valid JWT token.
2. A malicious attacker can make a user listing. You should not expose if a user exists in your database.